src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,260 @@
+
+    int64_t GetTimeoutMs() const { return timeout_ms_; }
+
+   private:
+    RetryableGrpcRequest(
+        std::function<void(std::shared_ptr<RetryableGrpcRequest> request)> executor,
+        std::function<void(ray::Status)> failure_callback,
+        size_t request_bytes,
+        int64_t timeout_ms)
+        : executor_(std::move(executor)),
+          failure_callback_(std::move(failure_callback)),
+          request_bytes_(request_bytes),
+          timeout_ms_(timeout_ms) {}
+
+    std::function<void(std::shared_ptr<RetryableGrpcRequest> request)> executor_;
+    std::function<void(ray::Status)> failure_callback_;
+    const size_t request_bytes_;
+    const int64_t timeout_ms_;
+  };
+
+ public:
+  static std::shared_ptr<RetryableGrpcClient> Create(
+      std::shared_ptr<grpc::Channel> channel,
+      instrumented_io_context &io_context,
+      uint64_t max_pending_requests_bytes,
+      uint64_t check_channel_status_interval_milliseconds,
+      uint64_t server_unavailable_timeout_seconds,
+      std::function<void()> server_unavailable_timeout_callback,
+      const std::string &server_name) {
+    return std::shared_ptr<RetryableGrpcClient>(
+        new RetryableGrpcClient(channel,
+                                io_context,
+                                max_pending_requests_bytes,
+                                check_channel_status_interval_milliseconds,
+                                server_unavailable_timeout_seconds,
+                                server_unavailable_timeout_callback,
+                                server_name));
+  }
+
+  RetryableGrpcClient(const RetryableGrpcClient &) = delete;
+  RetryableGrpcClient &operator=(const RetryableGrpcClient &) = delete;
+
+  template <typename Service, typename Request, typename Reply>
+  void CallMethod(PrepareAsyncFunction<Service, Request, Reply> prepare_async_function,
+                  std::shared_ptr<GrpcClient<Service>> grpc_client,
+                  const std::string &call_name,
+                  const Request &request,
+                  const ClientCallback<Reply> &callback,
+                  const int64_t timeout_ms);
+
+  void Retry(std::shared_ptr<RetryableGrpcRequest> request);
+
+  // Return the number of pending requests waiting for retry.
+  size_t NumPendingRequests() const { return pending_requests_.size(); }
+
+  ~RetryableGrpcClient();
+
+ private:
+  RetryableGrpcClient(std::shared_ptr<grpc::Channel> channel,
+                      instrumented_io_context &io_context,
+                      uint64_t max_pending_requests_bytes,
+                      uint64_t check_channel_status_interval_milliseconds,
+                      uint64_t server_unavailable_timeout_seconds,
+                      std::function<void()> server_unavailable_timeout_callback,
+                      const std::string &server_name)
+      : io_context_(io_context),
+        timer_(std::make_unique<boost::asio::deadline_timer>(io_context)),
```

Comment:
I'm not sure why do we use unique pointer for timer, and how plain type doesn't work here?